artifacts.go

```diff
@@ -311,14 +312,203 @@ func validateArtifact(c *cli.Context) error {
 	}
 
 	if !valid {
-		fmt.Println("Artifact file '" + c.Args().First() +
-			"' formatted correctly, but error validating signature.")
+		return cli.NewExitError("Artifact file '"+c.Args().First()+
+			"' formatted correctly, but error validating signature.", 1)
 	} else {
 		fmt.Println("Artifact file '" + c.Args().First() + "' validated successfully")
 	}
 	return nil
 }
 
+func createSignature(raw *artifact.Raw, w *awriter.Writer, key []byte) error {
+	s := artifact.NewSigner(key)
+	buf := bytes.NewBuffer(nil)
+	_, err := io.Copy(buf, raw.Data)
+	if err != nil {
+		return errors.Wrap(err, "Can not copy manifest data for signing")
+	}
+	signed, sErr := s.Sign(buf.Bytes())
+	if sErr != nil {
+		return sErr
+	}
+
+	// first write orifinal "manifest" file
+	if err :=
+		w.WriteRaw(artifact.NewRaw("manifest", raw.Size, buf)); err != nil {
+		return err
+	}
+	// then, write "manifest.sig"
+	if err :=
+		w.WriteRaw(
+			artifact.NewRaw("manifest.sig",
+				int64(len(signed)), bytes.NewBuffer(signed))); err != nil {
+		return err
+	}
+	return nil
+}
+
+func processHeader(r *areader.Reader, w *awriter.Writer,
+	key []byte, force bool) error {
+	// simple list with the header element name and the flag uses as
+	// indicator if element is optional or required
+	artifactHeaderElems := []struct {
+		name     string
+		required bool
+	}{
+		{"manifest", true},
+		{"manifest.sig", false},
+		{"header.", true},
+	}
+
+	getNext := true
+	var raw *artifact.Raw = nil
+	var err error = nil
+
+	// read header elements first
+	for _, elem := range artifactHeaderElems {
+		// get element form the artifact
+		if getNext {
+			raw, err = r.ReadRaw()
+			if err != nil {
+				return err
+			}
+		}
+		getNext = true
+
+		// check if we are not having element out of order
+		if elem.required && !strings.HasPrefix(raw.Name, elem.name) {
```

Comment:
This is basically what we are doing in all other places. This is in case we will change the compression algorithm from gz to something else.